File: stripe/src/main/java/com/stripe/android/view/PaymentUtils.java
package com.stripe.android.view;

import android.support.annotation.NonNull;

import java.text.DecimalFormatSymbols;
import java.text.NumberFormat;
import java.util.Currency;
import java.util.Locale;

public class PaymentUtils {

    /**
     * Formats a monetary amount into a human friendly string where zero is returned
     * as free.
     */
    static String formatPriceStringUsingFree(long amount, @NonNull Currency currency, String free) {
        if (amount == 0) {
            return free;
        }
        NumberFormat currencyFormat = NumberFormat.getCurrencyInstance();
        DecimalFormatSymbols decimalFormatSymbols = ((java.text.DecimalFormat) currencyFormat)
                .getDecimalFormatSymbols();
        decimalFormatSymbols.setCurrencySymbol(currency.getSymbol(Locale.getDefault()));
        ((java.text.DecimalFormat) currencyFormat).setDecimalFormatSymbols(decimalFormatSymbols);

        return formatPriceString(amount, currency);
    }

    /**
     * Formats a monetary amount into a human friendly string.
     */
    static String formatPriceString(double amount, @NonNull Currency currency) {
        double majorUnitAmount = amount / Math.pow(10, currency.getDefaultFractionDigits());
        NumberFormat currencyFormat = NumberFormat.getCurrencyInstance();
        try {
            DecimalFormatSymbols decimalFormatSymbols = ((java.text.DecimalFormat) currencyFormat)
                    .getDecimalFormatSymbols();
            decimalFormatSymbols.setCurrencySymbol(currency.getSymbol(Locale.getDefault()));
            ((java.text.DecimalFormat) currencyFormat)
                    .setDecimalFormatSymbols(decimalFormatSymbols);
            return currencyFormat.format(majorUnitAmount);
        } catch (ClassCastException e) {
            return currencyFormat.format(majorUnitAmount);
        }
    }

}

File: payments-core/src/main/java/com/stripe/android/view/PaymentUtils.kt
package com.stripe.android.view

import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.Currency
import java.util.Locale
import kotlin.math.pow

object PaymentUtils {

    /**
     * Formats a monetary amount into a human friendly string where zero is returned
     * as free.
     */
    @JvmStatic
    fun formatPriceStringUsingFree(
        amount: Long,
        currency: Currency,
        free: String
    ): String {
        if (amount == 0L) {
            return free
        }

        val currencyFormat = NumberFormat.getCurrencyInstance()
        val decimalFormatSymbols = (currencyFormat as DecimalFormat)
            .decimalFormatSymbols
        decimalFormatSymbols.currencySymbol = currency.getSymbol(Locale.getDefault())
        currencyFormat.decimalFormatSymbols = decimalFormatSymbols

        return formatPriceString(amount.toDouble(), currency)
    }

    /**
     * Formats a monetary amount into a human friendly string.
     */
    @JvmSynthetic
    internal fun formatPriceString(amount: Double, currency: Currency): String {
        val majorUnitAmount = amount / 10.0.pow(currency.defaultFractionDigits.toDouble())
        val currencyFormat = NumberFormat.getCurrencyInstance()
        return try {
            val decimalFormatSymbols = (currencyFormat as DecimalFormat)
                .decimalFormatSymbols
            decimalFormatSymbols.currencySymbol = currency.getSymbol(Locale.getDefault())
            currencyFormat.decimalFormatSymbols = decimalFormatSymbols
            currencyFormat.format(majorUnitAmount)
        } catch (e: ClassCastException) {
            currencyFormat.format(majorUnitAmount)
        }
    }
}
